sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -1517,6 +2231,99 @@ submitted in its entirety for execution via
   ...
 ----
 
+=== Dynamic Parameter Update
+
+Example showing a graph with two kernel nodes, the first created using a kernel
+bundle with `handler::set_args()`, having its node arguments updated.
+
+[source,c++]
+----
+...
+
+using namespace sycl;
+namespace sycl_ext = sycl::ext::oneapi::experimental;
+
+queue myQueue;
+auto myContext = myQueue.get_context();
+auto myDevice = myQueue.get_device();
+
+// USM allocations for kernel input/output
+const size_t n = 1024;
+int *ptrX = malloc_shared<int>(n, myQueue);
+int *ptrY = malloc_device<int>(n, myQueue);
+
+int *ptrZ = malloc_shared<int>(n, myQueue);
+int *ptrQ = malloc_device<int>(n, myQueue);
+
+// Initialize USM allocations
+
+// Kernel loaded from kernel bundle
+const std::vector<kernel_id> builtinKernelIds =
+      myDevice.get_info<info::device::built_in_kernel_ids>();
+kernel_bundle<bundle_state::executable> myBundle =
+      get_kernel_bundle(myContext, { myDevice }, builtinKernelIds);
+kernel builtinKernel = myBundle.get_kernel(builtinKernelIds[0]);
+
+// Graph containing a two kernels node
+sycl_ext::command_graph myGraph(myContext, myDevice);
+
+// Create graph dynamic parameters
+dynamic_parameter<sycl_ext::dynamic_parameter_type::usm> dynParamInput(myGraph);
+dynamic_parameter<sycl_ext::dynamic_parameter_type::usm> dynParamOutput(myGraph);
+dynamic_parameter<sycl_ext::dynamic_parameter_type::scalar> dynParamScalar(myGraph);
+
+// First node uses ptrX as an input & output parameter, with operand
+// mySclar as another argument.
+int myScalar = 42;
+node nodeA = myGraph.add[&](handler& cgh) {
+    cgh.set_args(ptrX, myScalar);
+    cgh.parallel_for(range {n}, builtinKernel);
+});
+
+// Register nodeA dynamic parameters
+dynParamInput.register(0, nodeA);  // Argument index 0 is ptrX
+dynParamInput.register(1, nodeA);  // Argument index 1 is myScalar
+
+// Second node uses ptrX as an inputs, and ptrY as input/output/
+node nodeB = myGraph.add[&](handler& cgh) {
+    cgh.parallel_for(sycl::range<1>{n}, [=](sycl::id<1> it) {
+      const size_t i = it[0];
+      ptrY[i] += ptrX[i];
+    });
+   },
+   {sycl_ext::property::node::depends_on(nodeA)});
+
+// Register nodeB dynamic parameters
+dynParamInput.register(ptrX, nodeB);
+dynParamInput.register(ptrY, nodeB);
+
```

Comment:
Adding another comment here to keep this conversation separately threaded.

I don't understand how `dynamic_parameter::update` works for an accessor.  How would you create the new accessor value?  Normally, an accessor is created inside of command group scope, but there is no command group when you update the parameter values.  Are you expecting that the user creates a placeholder accessor?

I'm also not clear on how parameter update works for a recorded graph.  Is this why you added `command_graph::get_nodes` and `command_graph::get_root_nodes`?  Is the idea that the application uses these to get specific nodes in the graph and then it figures out the parameters to those nodes by somehow knowing the values of the variables that are captured in that node?  This seems to require a lot of knowledge sharing between the code that constructs the `dynamic_parameter` and the code that submits commands to the queue that is being recorded.

Reply:
> how parameter update works for a recorded graph

for the USM update, see https://github.com/reble/llvm/pull/340/files#r1422564811

Reply:
Yes, I saw that comment.  Updating a USM pointer seems clear to me.  My question is specifically about updating an accessor.

Reply:
> I don't understand how `dynamic_parameter::update` works for an accessor. How would you create the new accessor value? Normally, an accessor is created inside of command group scope, but there is no command group when you update the parameter values. Are you expecting that the user creates a placeholder accessor?

Yes as it stands I think this would require placeholder accessors. An alternative could be to instead use the actual buffer when updating the parameter, so in the kernel the accessor would have the same properties but point to updated memory. Accessors may allow more flexibility in terms of updating the range or offset, but it's not clear to me how much that would be useful - it seems like most use cases would want the access to stay the same but simply swap out the actual memory between executions. However, changing offsets could be achieved with sub-buffers potentially which would restore some flexibility.

> I'm also not clear on how parameter update works for a recorded graph. Is this why you added `command_graph::get_nodes` and `command_graph::get_root_nodes`? Is the idea that the application uses these to get specific nodes in the graph and then it figures out the parameters to those nodes by somehow knowing the values of the variables that are captured in that node? This seems to require a lot of knowledge sharing between the code that constructs the `dynamic_parameter` and the code that submits commands to the queue that is being recorded.

This is a partial motivation yes, and it does require some knowledge sharing as you say. In the kinds of cases we are looking at though where memory is provided by the user to a library to perform some operations that knowledge is already present in the user code, and the queries allow the user access to what was recorded to be able to update those inputs. 

More broadly though these queries enable better mixed record & replay/explicit API usage by allowing nodes which have been recorded to be used as dependencies in the explicit API. Previously this mixed functionality was possible but only in a very limited way.

Reply:
also curious about how the accessor is implemented, is it a struct that contains size, offset, format (for image), memory handle (to buffer/image) etc.?

Reply:
looks that we need to add command_graph::get_nodes back, in framework, it is used to check if the recorded graph is empty, see at https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/cuda/CUDAGraph.cpp#L227

Reply:
Support for query functions like `get_nodes` can be found in a separate PR now: https://github.com/reble/llvm/pull/348